### PR TITLE
3 disentangle data processing from recording class of birdnetlib

### PR DESCRIPTION
**Pull request by MaHaWo**
- [x] build a separate `Preprocessor` class that takes the data processing steps originally contained in `birdnetlib.Recording`. 
- [x] add ABC for preprocessor to make interface clear
- [x] build a new  `SparrowAnalyzer` class to get custom birdnet-classifiers to work. This *might* be temporary and be replaced later, at least in name. See [here](https://github.com/joeweiss/birdnetlib/issues/108) for why it's necessary.
- [x] build a new `SparrowRecording` child class of `birdnetlib.Recording` that uses the preprocessor. 
- [x] use YAML for providing variables/file paths etc. -> maybe use for default config storage later? 
- [x] add custom test fixtures in separate files, folder.  
- [x] add tests for recording 
- [x] add tests for perprocessor of default model (we need to ship that)
- [x] add tests for analyzer   
- [x] add test sound files via lfs
- [x] add mock installation via pytest fixture with session scope 
- [x] remove lfs in workflow again.

### CODE REVIEW

**Comment by MaHaWo** [comment]
@iulusoy I think this one is ready for review if you have time. It looks a little daunting wrt to size, though most of the added stuff are tests that should be relatvely straight forward. 
Will try to limit future PR sizes.

- There's a test case for thrown exceptions in  that is currently commented out at the end of preprocessor_test.py, because each operating system threw a different one. Could replace the code that does the exception handling (copied over from birdnetlib) to do something generic, but that would be uninformative... 
- A bunch of deprecation warnings are thrown upon test execution - these stem mostly from the librosa library for audio processing, not much we can do about it I suppose...

File: models/google_perch/preprocessor.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
+import numpy as np
+import librosa
+import audioread
+from birdnetlib.exceptions import AudioFormatError
+
+from iSparrow import preprocessor_base as ppb
+
```
Comment (MaHaWo) [inline models/google_perch/preprocessor.py:7]: This will be filled in in one of the next PRs, as it's not directly compatible with birdnetlib.

**Review by iulusoy — COMMENTED** [review]
Ran the pre-commit hooks for the project, thus the changes - I would add flake8 to the hooks. 
```
https://flake8.pycqa.org/en/latest/user/using-hooks.html#usage-with-the-pre-commit-git-hooks-framework
```
`pooch` is great for generic downloads, but Hugging Face also has dedicated libraries. For now, for the model files, I would leave it as is, but in the longer term each model needs its own repo and should then be downloaded using the Hugging Face API.
Same holds for data, each dataset has its own dataset repo and would then be downloaded using the `datasets` library. It is dedicated to this purpose and also handles the feature assignment correctly.
The tests currently do not run on my machine, since the `install_config.yml` is missing. I think that is after the model download.
Otherwise looks good so far!

**Comment by iulusoy** [comment]
Also added sonarcloud analysis of the code quality; `flake8` as pre-commit hook can spot some bugs, and sonarcloud finds some too occasionally. It also is more strict with raising issues when not adhering to Python conventions.

File: requirements-dev.txt [hunk 2]
```diff
@@ -1,3 +1,6 @@
 sphinx
 myst-parser
-pytest
\ No newline at end of file
+pytest
+pytest-cov 
+coverage
+pooch
```
[the remark this answers is not in the document]
Comment (MaHaWo) [reply requirements-dev.txt]: removed the `pooch` one.

File: tests/conftest.py [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+    label_file_names = {
+        "birdnet_default_v2.4/labels.txt": None,
+        "birdnet_custom_v2.4/labels.txt": None,
+        "google_bird_classification/assets/family.csv": None,
+        "google_bird_classification/assets/genus.csv": None,
+        "google_bird_classification/assets/label.csv": None,
+        "google_bird_classification/assets/order.csv": None,
+    }
+
+    models_data = pooch.create(
```
[the remark this answers is not in the document]
Comment (MaHaWo) [reply tests/conftest.py:113]: yes, better. wasn't aware of the `huggingface_hub`.

File: models/birdnet_default/preprocessor.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+from src.iSparrow import preprocessor_base as ppb
+
+
+class Preprocessor(ppb.PreprocessorBase):
+    """
+    Preprocessor Preprocess audio data into resampled chunks for analysis.
+
+    """
+
+    def __init__(self, sample_rate: int = 48000, overlap: float = 0.0, sample_secs: int = 3.0, resample_type: str = "kaiser_fast"):
```
[the remark this answers is not in the document]
Comment (MaHaWo) [reply models/birdnet_default/preprocessor.py]: ... was missin in the repo on my notebook

**Comment by MaHaWo** [comment]
thanks for the late-night review! 

- the tests should run now - re-added the config that got lost along the way. 
- `hugginface_hub` sounds good. However, given that this took me a lot longer than I had hoped, I would propose to give that a lower priority for the time being.

File: models/birdnet_default/preprocessor.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (MaHaWo) [reply models/birdnet_default/preprocessor.py]: It was missing on my laptop.... Office machine should have it.

**Pull request merged**